### PR TITLE
Fix load_state_dict_post_hooks BC

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77281

Certain modules (such as `nn.Softmax`) may not dispatch into __setstate__ of nn.Module, so `_load_state_dict_post_hooks` might be undefined, causing errors.

As a stopgap, fix this by checking if the attr exists on the module before accessing it. Also, create the attribute if it does not exist when registering these hooks, instead of assuming it will exist.

The proper fix is actually to fix https://github.com/pytorch/pytorch/issues/77280 and ensure nn.Module instances dispatch into superclass of __setstate__. We should probably also add documentation that if users override their own nn.Module setstate method, it needs to dispatch into the superclass, otherwise there will be BC issues.

More context in https://github.com/pytorch/pytorch/issues/77280.

Differential Revision: [D36321915](https://our.internmc.facebook.com/intern/diff/D36321915/)

### CODE REVIEW

**Review by jbschlosser — CHANGES_REQUESTED** [review]
As discussed externally, the fix of calling `super().__setstate__()` from the overrides for `nn.Softmax`, `nn.Softmin`, and `nn.LogSoftmax` is preferred. Test logic looks nice and should still apply though :)
